pkg/daemon/update.go

```diff
@@ -1073,55 +1106,63 @@ func (dn *Daemon) deleteStaleData(oldIgnConfig, newIgnConfig *ign3types.Config)
-				var restore bool
-				if _, err := exec.Command("rpm", "-qf", f.Path).CombinedOutput(); err == nil {
-					// File is owned by an rpm
-					restore = true
-				} else if strings.HasPrefix(f.Path, "/etc") && (operatingSystem == MachineConfigDaemonOSRHCOS || operatingSystem == MachineConfigDaemonOSFCOS) {
-					if _, err := os.Stat("/usr" + f.Path); err != nil {
-						if !os.IsNotExist(err) {
-							return err
-						}
-
-						// If the error is ErrNotExist then we don't restore the file
-					} else {
-						restore = true
-					}
-				}
-
-				if restore {
-					if err := restorePath(f.Path); err != nil {
+		if _, ok := newFileSet[f.Path]; ok {
+			continue
+		}
+		if _, err := os.Stat(noOrigFileStampName(f.Path)); err == nil {
+			if err := os.Remove(noOrigFileStampName(f.Path)); err != nil {
+				return errors.Wrapf(err, "deleting noorig file stamp %q: %v", noOrigFileStampName(f.Path), err)
+			}
+			glog.V(2).Infof("Removing file %q completely", f.Path)
+		} else if _, err := os.Stat(origFileName(f.Path)); err == nil {
+			// Add a check for backwards compatibility: basically if the file doesn't exist in /usr/etc (on FCOS/RHCOS)
+			// and no rpm is claiming it, we assume that the orig file came from a wrongful backup of a MachineConfig
+			// file instead of a file originally on disk. See https://bugzilla.redhat.com/show_bug.cgi?id=1814397
+			var restore bool
+			if _, err := exec.Command("rpm", "-qf", f.Path).CombinedOutput(); err == nil {
+				// File is owned by an rpm
+				restore = true
+			} else if strings.HasPrefix(f.Path, "/etc") && (operatingSystem == MachineConfigDaemonOSRHCOS || operatingSystem == MachineConfigDaemonOSFCOS) {
+				if _, err := os.Stat("/usr" + f.Path); err != nil {
+					if !os.IsNotExist(err) {
 						return err
 					}
-					glog.V(2).Infof("Restored file %q", f.Path)
-					continue
-				}
 
-				if err := os.Remove(origFileName(f.Path)); err != nil {
-					return errors.Wrapf(err, "deleting orig file %q: %v", origFileName(f.Path), err)
+					// If the error is ErrNotExist then we don't restore the file
+				} else {
+					restore = true
 				}
 			}
-			glog.V(2).Infof("Deleting stale config file: %s", f.Path)
-			if err := os.Remove(f.Path); err != nil {
-				newErr := fmt.Errorf("unable to delete %s: %s", f.Path, err)
-				if !os.IsNotExist(err) {
-					return newErr
+
+			if restore {
+				if err := restorePath(f.Path); err != nil {
+					return err
 				}
-				// otherwise, just warn
-				glog.Warningf("%v", newErr)
+				glog.V(2).Infof("Restored file %q", f.Path)
+				continue
+			}
+
+			if err := os.Remove(origFileName(f.Path)); err != nil {
+				return errors.Wrapf(err, "deleting orig file %q: %v", origFileName(f.Path), err)
+			}
+		}
+
+		// Check Systemd.Units.Dropins - don't remove the file if configuration has been converted into a dropin
+		if dn.isPathInDropins(f.Path, &newIgnConfig.Systemd) {
+			glog.V(2).Infof("Not removing file %q: replaced with systemd dropin", f.Path)
```

Comment:
Probably should just log with default verbosity here. Shouldn't happen very often and would be nice for us to know if this causes some edge case in the future

Reply:
Agree, using the same verbosity as in `Removed stale file` message